src/Protocol/Diaspora.php

```diff
@@ -1414,6 +1414,41 @@ private static function message($guid, $server, $level = 0)
 		return $msg;
 	}
 
+	/**
+	 * @brief Fetches an item with a given URL
+	 *
+	 * @param string $url the message url
+	 *
+	 * @return int the message id of the stored message or false
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 * @throws \ImagickException
+	 */
+	public static function fetchByURL($url, $uid = 0)
+	{
+		if (!preg_match("=([http|https].*)/(.*)/(.*)=ism", $url, $matches)) {
```

Comment:
If all Diaspora post URLs are of the form `http(s)://example.com/{post|display}/{guid}` then this regular expression can and probably should be more specific than that. For example, this condition matches the string `|//` which probably isn't what you expected.

I suggest `(https?://.+)/(?:posts|display)/([a-f0-9]{32})` instead, you can remove the condition on line 1433 and all subsequent `$matches[3]` become `$matches[2]`.

Reply:
The guid can contain of any characters of any length. I'm totally confused by ```(?:posts|display)``` as well. Shouldn't it be ```[posts|display].*``` instead? I'm really not deep into this regexp stuff.

Reply:
Any character, are you sure? There's absolutely no limitation?

As for the difference between  and `[posts|display].*`:
- `(?:posts|display)` means "either `posts` or `display`" in a non-capturing group (won't appear in `$matches`)
- `[posts|display].*` means "one character in the list `p, o, s, t, |, d, i, l, a, y` followed by any number of characters". Square brackets define the allowed characters, not literal strings. Inside square brackets, most special characters (like `|` and `?) are treated like literal characters to be expected in the subject.

Reply:
Diaspora GUID specification:

**GUID**

A network-wide, unique identifier. A random string of at least 16 and at most 255 chars. It contains only:

```
    Letters: a-z and A-Z
    Numbers: 0-9
    Special chars: -, _, @, . and :
```

Special chars aren’t allowed at the end.

Reply:
Revised regular expression: `=(https?://.+)/(?:posts|display)/([a-zA-Z0-9-_@.:]+[a-zA-Z0-9])=i`. I also removed the extraneous flags.

Reply:
Well, there is some more exception. Hubzilla uses the ```@``` - but since this is an URL, it is encoded as ```%40```. (Reminds me of urldecoding the guid before parsing it to the function).

Reply:
Updated regular expression: `=(https?://.+)/(?:posts|display)/([a-zA-Z0-9-_@.:%]+[a-zA-Z0-9])=i`